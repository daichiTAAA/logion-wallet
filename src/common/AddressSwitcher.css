.AddressSwitcher {
    height: 75px;
    width: 430px;
    display: flex;
}

.AddressSwitcher .dropdown {
    width: 100%;
}

#address-switcher-toggle {
    width: 100%;
    height: 75px;
    background-color: transparent;
    margin: 0;
    padding: 0;
    overflow: hidden;
    border: none;
    border-radius: 0;
    text-align: left;
}

#address-switcher-toggle:focus {
    box-shadow: none;
}

#address-switcher-toggle .address-data {
    display: inline-block;
    width: calc(100% - 20px);
}

.AddressSwitcher .dropdown-menu {
    width: 100%;
    padding: 0;
    margin: 0;
}

.AddressSwitcher .dropdown-menu .dropdown-item {
    padding-top: 10px;
    padding-bottom: 10px;
    padding-left: 0;
    padding-right: 0;
    margin: 0;
}

.AddressSwitcher .dropdown-menu .dropdown-item:focus {
    background-color: transparent;
}
