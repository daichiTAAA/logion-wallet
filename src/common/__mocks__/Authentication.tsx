export let authenticate = jest.fn();
