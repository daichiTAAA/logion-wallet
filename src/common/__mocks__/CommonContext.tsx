import { 
    useCommonContext as useCommonContextMock
} from './CommonContextMock';

export const useCommonContext = useCommonContextMock;
