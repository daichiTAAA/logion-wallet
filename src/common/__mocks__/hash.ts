import { sha256Hex } from "./HashMock";
export { sha256Hex }
