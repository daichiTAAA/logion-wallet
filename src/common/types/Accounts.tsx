import moment, { Moment } from 'moment';
import { InjectedAccountWithMeta } from '@polkadot/extension-inject/types';

export interface Token {
    readonly value: string;
    readonly expirationDateTime: Moment;
}

export interface Account {
    readonly name: string,
    readonly address: string,
    readonly isLegalOfficer: boolean,
    readonly token?: Token,
}

export default interface Accounts {
    readonly all: Account[],
    readonly current?: Account,
}

export class AccountTokens {

    constructor(initialState: Record<string, Token>) {
        this.store = { ...initialState };
    }

    private store: Record<string, Token>;

    get(address: string): Token | undefined {
        return this.store[address];
    }

    merge(tokens: AccountTokens): AccountTokens {
        const newStore = { ...this.store };
        for(const address of tokens.addresses) {
            newStore[address] = tokens.store[address];
        }
        return new AccountTokens(newStore);
    }

    get addresses(): string[] {
        return Object.keys(this.store);
    }

    cleanUp(now: Moment): AccountTokens {
        const newStore: Record<string, Token> = {};
        for(const address of this.addresses) {
            const token = this.get(address)!;
            if(token.expirationDateTime.isAfter(now)) {
                newStore[address] = token;
            }
        }
        return new AccountTokens(newStore);
    }

    equals(other: AccountTokens): boolean {
        if(this.length !== other.length) {
            return false;
        }
        for(const address of this.addresses) {
            const thisToken = this.get(address);
            const otherToken = other.get(address);
            if(thisToken!.value !== otherToken?.value
                || !thisToken!.expirationDateTime.isSame(otherToken.expirationDateTime)) {
                return false;
            }
        }
        return true;
    }

    get length(): number {
        return this.addresses.length;
    }

    isAuthenticated(now: Moment, address: string | undefined): boolean {
        if(address === undefined) {
            return false;
        }
        const token = this.get(address);
        if(token === undefined) {
            return false;
        } else {
            return token.expirationDateTime.isAfter(now);
        }
    }

    earliestExpiration(): Moment | undefined {
        let earliest: Moment | undefined;
        for(const address of this.addresses) {
            const expiration = this.store[address].expirationDateTime;
            if(earliest === undefined || earliest.isAfter(expiration)) {
                earliest = expiration;
            }
        }
        return earliest;
    }
}

export function buildAccounts(
    injectedAccounts: InjectedAccountWithMeta[],
    userAddress: string | undefined,
    tokens: AccountTokens,
    isLegalOfficer: (address: string) => boolean,
): Accounts {
    const selectedAddress = currentOrDefaultAddress(injectedAccounts, userAddress, tokens);

    let current: Account | undefined;
    const matchingAddress = injectedAccounts.find(injectedAccount => injectedAccount.address === selectedAddress);
    if(selectedAddress !== undefined && matchingAddress !== undefined) {
        current = {
            name: matchingAddress.meta.name!,
            address: selectedAddress!,
            isLegalOfficer: isLegalOfficer(selectedAddress),
            token: tokenOrUndefinedIfExpired(tokens.get(selectedAddress)),
        }
    } else {
        current = undefined;
    }

    return {
        current,
        all: injectedAccounts.map(injectedAccount => ({
            name: injectedAccount.meta.name!,
            address: injectedAccount.address,
            isLegalOfficer: isLegalOfficer(injectedAccount.address),
            token: tokenOrUndefinedIfExpired(tokens.get(injectedAccount.address)),
        }))
    }
}

function currentOrDefaultAddress(
    injectedAccounts: InjectedAccountWithMeta[],
    currentAddress: string | undefined,
    tokens: AccountTokens
): string | undefined {
    const loggedAddresses = tokens.addresses;
    if(currentAddress !== undefined) {
        return currentAddress;
    } else if(loggedAddresses.length > 0) {
        const defaultAccount = injectedAccounts.find(account => loggedAddresses.includes(account.address));
        return defaultAccount?.address;
    } else {
        return undefined;
    }
}

function tokenOrUndefinedIfExpired(token: Token | undefined): Token | undefined {
    if(token === undefined) {
        return undefined;
    }

    const expirationDateTime = token.expirationDateTime;
    if(expirationDateTime.isBefore(moment())) {
        return undefined;
    } else {
        return token;
    }
}
