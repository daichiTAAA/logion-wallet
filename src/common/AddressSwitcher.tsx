import React from 'react';
import Dropdown from 'react-bootstrap/Dropdown';

import Addresses from './types/Addresses';

import './AddressSwitcher.css';
import AccountAddress from './AccountAddress';
import { useCommonContext } from './CommonContext';

export interface Props {
    addresses: Addresses | null,
    selectAddress: ((userAddress: string) => void) | null,
}

export default function AddressSwitcher(props: Props) {
    const { colorTheme } = useCommonContext();

    if(props.addresses === null || props.selectAddress === null) {
        return null;
    }

    return (
        <div
            className="AddressSwitcher"
            style={{
                color: colorTheme.accounts.foreground
            }}
        >
            <Dropdown>
                <Dropdown.Toggle id="address-switcher-toggle">
                    <div className="address-data">
                        <AccountAddress
                            hint="Click to select another address"
                            address={ props.addresses.currentAddress }
                        />
                    </div>
                </Dropdown.Toggle>
                <Dropdown.Menu>
                    {
                        props.addresses.addresses.map(address => (
                            <Dropdown.Item
                                key={ address.address }
                                onClick={ () => props.selectAddress!(address.address) }
                                style={{
                                    color: colorTheme.accounts.foreground,
                                    backgroundColor: colorTheme.accounts.background,
                                }}
                            >
                                <AccountAddress
                                    address={ address }
                                />
                            </Dropdown.Item>
                        ))
                    }
                </Dropdown.Menu>
            </Dropdown>
        </div>
    );
}
