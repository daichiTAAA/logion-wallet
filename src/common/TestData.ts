import { Coin, CoinBalance, SYMBOL } from '../logion-chain/Balances';
import { PrefixedNumber, ATTO } from '../logion-chain/numbers';

import Identity from './types/Identity';
import PostalAddress from './types/PostalAddress';
import { Transaction } from './types/ModelTypes';
import { ColorTheme, rgbaToHex } from './ColorTheme';

export const DEFAULT_LEGAL_OFFICER = "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"; // Alice
export const ANOTHER_LEGAL_OFFICER = "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty"; // Bob
export const A_THIRD_LEGAL_OFFICER = "5FLSigC9HGRKVhB9FiEo4Y3koPsNmBmLJbpXg2mp1hXcS59Y"; // Charlie

export const DEFAULT_IDENTITY: Identity = {
    firstName: "John",
    lastName: "Doe",
    email: "john.doe@logion.network",
    phoneNumber: "+1234",
};

export const DEFAULT_ADDRESS: PostalAddress = {
    line1: "Place de le République Française, 10",
    line2: "boite 15",
    postalCode: "4000",
    city: "Liège",
    country: "Belgium",
};

export const COLOR_THEME: ColorTheme = {
    type: 'light',
    shadowColor: rgbaToHex('#3b6cf4', 0.1),
    dashboard: {
        background: '#eff3fe',
        foreground: '#000000',
    },
    sidebar: {
        background: '#ffffff',
        foreground: '#000000',
        activeItemBackground: '#d8e2fd',
    },
    accounts: {
        iconBackground: '#3b6cf4',
        hintColor: rgbaToHex('#000000', 0.6),
        foreground: '#000000',
        background: 'eff3fe',
        legalOfficerIcon: {
            category: 'legal-officer',
            id: 'account-shield'
        }
    },
    frame: {
        background: '#ffffff',
        foreground: '#000000',
        link: rgbaToHex('#3b6cf4', 0.15),
        altBackground: '#eff3fe',
    },
    topMenuItems: {
        iconGradient: {
            from: '#3b6cf4',
            to: '#6050dc',
        }
    },
    bottomMenuItems: {
        iconGradient: {
            from: '#7a90cb',
            to: '#3b6cf4',
        }
    },
    recoveryItems: {
        iconGradient: {
            from: '#7a90cb',
            to: '#3b6cf4',
        }
    },
    buttons: {
        secondary: {
            background: '#eff3fe',
            foreground: '#000000',
        }
    },
    select: {
        background: '#ffffff',
        foreground: '#000000',
        menuBackgroundColor: '#ffffff',
        selectedOptionBackgroundColor: '#ffffff',
    },
    table: {
        background: '#0c163d',
        foreground: '#ffffff',
        header: {
            background: '#0c163d',
            foreground: '#ffffff',
        },
        row: {
            background: '#152665',
            foreground: '#ffffff',
        }
    },
    tabs: {
        background: '#0c163d',
        foreground: '#ffffff',
        inactiveTab: {
            background: '#d8e2fd',
            foreground: '#ffffff',
        },
        borderColor: '#3b6cf4',
    },
    dialog: {
        background: '#eff3fe',
        foreground: '#000000',
        borderColor: '#e6007a',
    },
};

export const DEFAULT_COIN: Coin = {
    id: 'log',
    iconId: 'log',
    iconType: 'png',
    name: "Logion",
    symbol: SYMBOL
};

export const DEFAULT_COIN_BALANCE: CoinBalance = {
    coin: DEFAULT_COIN,
    balance: new PrefixedNumber("42", ATTO),
    level: 0.1,
};

export const DEFAULT_TRANSACTION: Transaction = {
    from: "from",
    to: "to",
    pallet: "pallet",
    method: "method",
    transferValue: "420",
    tip: "0",
    fee: "125000",
    reserved: "0",
    total: "125420",
    createdOn: "2021-07-28T12:30:00.000",
    type: "Sent",
}
