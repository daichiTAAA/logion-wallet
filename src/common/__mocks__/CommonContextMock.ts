import { TEST_WALLET_USER } from '../../wallet-user/TestData';
import { COLOR_THEME } from '../TestData';
import Addresses, { AccountAddress } from '../types/Addresses';

export let selectAddress = jest.fn();

export let currentAddress = TEST_WALLET_USER;

export function setCurrentAddress(address: string) {
    currentAddress = address;
}

const DEFAULT_ACCOUNT: AccountAddress = {
    name: "name",
    address: TEST_WALLET_USER,
    isLegalOfficer: false,
}

export let addresses: Addresses = {
    currentAddress: DEFAULT_ACCOUNT,
    addresses: [ DEFAULT_ACCOUNT ]
}

export let balances: any = null;

export let transactions: any = null;

export function useCommonContext() {
    return {
        currentAddress,
        selectAddress,
        addresses,
        balances,
        transactions,
        colorTheme: COLOR_THEME,
    };
}

export function setAddresses(value: Addresses) {
    addresses = value;
}

export function setBalances(value: any) {
    balances = value;
}

export function setTransactions(value: any) {
    transactions = value;
}
