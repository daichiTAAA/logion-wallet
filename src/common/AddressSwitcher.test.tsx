import moment from 'moment';
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Accounts, { Account } from './types/Accounts';
import { shallowRender } from '../tests';
import AddressSwitcher from './AddressSwitcher';
import { setAddresses, authenticate } from './__mocks__/CommonContextMock';
import { TEST_WALLET_USER } from '../wallet-user/TestData';
import { DEFAULT_LEGAL_OFFICER } from "./TestData";

jest.mock('./CommonContext');
jest.mock('./Authentication');

const AUTHENTICATED_ADDRESS: Account = {
    name: "name authenticated",
    address: TEST_WALLET_USER,
    isLegalOfficer: false,
    token: {
        value: "token",
        expirationDateTime: moment('2021-09-10T10:53:00.000Z'),
    },
};

const UNAUTHENTICATED_ADDRESS: Account = {
    name: "name unauthenticated",
    address: DEFAULT_LEGAL_OFFICER,
    isLegalOfficer: true,
};

const ADDRESSES: Accounts = {
    current: AUTHENTICATED_ADDRESS,
    all: [ UNAUTHENTICATED_ADDRESS ]
};

describe("AddressSwitcher", () => {

    it("renders", () => {
        setAddresses(ADDRESSES);
        const result = shallowRender(
            <AddressSwitcher
                selectAddress={ () => {} }
            />
        );
        expect(result).toMatchSnapshot();
    });

    it("enables log-in given unautenticated yet user", async () => {
        setAddresses(ADDRESSES);
        render(
            <AddressSwitcher
                selectAddress={ () => {} }
            />
        );

        userEvent.click(screen.getByText("Click to select another address"));
        await waitFor(() => userEvent.click(screen.getByRole('button', {name:"login button"})));

        await waitFor(() => expect(authenticate).toBeCalledWith([ DEFAULT_LEGAL_OFFICER ]));
    });
});
