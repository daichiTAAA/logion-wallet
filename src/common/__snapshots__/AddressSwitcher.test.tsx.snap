// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressSwitcher renders 1`] = `
<div
  className="AddressSwitcher light"
  style={
    Object {
      "color": "#000000",
    }
  }
>
  <Dropdown
    navbar={false}
  >
    <DropdownToggle
      id="address-switcher-toggle"
    >
      <div
        className="address-data"
      >
        <AccountAddress
          address={
            Object {
              "address": "5H4MvAsobfZ6bBCDyj5dsrWYLrA8HrRzaqa9p61UXtxMhSCY",
              "isLegalOfficer": false,
              "name": "name authenticated",
              "token": Object {
                "expirationDateTime": "2021-09-10T10:53:00.000Z",
                "value": "token",
              },
            }
          }
          disabled={false}
          hint="Click to select another address"
        />
      </div>
    </DropdownToggle>
    <DropdownMenu
      align="left"
      alignRight={false}
      flip={true}
      style={
        Object {
          "backgroundColor": "eff3fe",
          "boxShadow": "0 0 25px #3b6cf41a",
          "color": "#000000",
        }
      }
    >
      <DropdownItem
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          }
        }
        disabled={false}
        onClick={[Function]}
      >
        <AccountAddress
          address={
            Object {
              "address": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
              "isLegalOfficer": true,
              "name": "name unauthenticated",
            }
          }
          disabled={true}
          login={[Function]}
        />
      </DropdownItem>
      <DropdownItem
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          }
        }
        className="logout"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "borderTop": "1px solid #000000",
          }
        }
      >
        <Button
          variant="salmon"
        >
          Log Out
        </Button>
      </DropdownItem>
    </DropdownMenu>
  </Dropdown>
  <Dialog
    actions={
      Array [
        Object {
          "buttonText": "Confirm",
          "buttonVariant": "primary",
          "callback": undefined,
          "id": "confirm",
        },
        Object {
          "buttonText": "Cancel",
          "buttonVariant": "secondary",
          "callback": [Function],
          "id": "cancel",
        },
      ]
    }
    show={false}
    size="lg"
  >
    <p>
      This action will log you off from 
      <strong>
        all
      </strong>
       current connected accounts, do you confirm ?
    </p>
  </Dialog>
</div>
`;
